### ISSUE TITLE
sysconfig_notcgo does not implement GetLongBit

### ISSUE THREAD

**Issue by hatchan** [opencontainers/runc#841]
When compiling with `CGO_ENABLED=0` this results in a build error. This occurs for us because this package is a dependency of a dependency, and we're building our package using `CGO_ENABLED=0`. The error is that `system.GetLongBit` is not defined.

This is where `GetLongBit` defined for most build: https://github.com/opencontainers/runc/blob/master/libcontainer/system/sysconfig.go#L29

This file is used when CGO is not enabled: https://github.com/opencontainers/runc/blob/master/libcontainer/system/sysconfig_notcgo.go

`GetLongBit` is only used in `libcontainer/cgroups/fs` so you can reproduce it as follows:

```
$ cd $GOPATH/src/github.com/opencontainers/runc/libcontainer/cgroups/fs
$ CGO_ENABLED=0 go build
# github.com/opencontainers/runc/libcontainer/cgroups/fs
./memory.go:71: undefined: system.GetLongBit
```

**Comment by cyphar** [comment]
Looks like we need to split `sysconfig.go` into `sysconfig.go` (with the truly agnostic stuff) and `sysconfig_cgo.go` for the cgo-specific things.

**Comment by crosbymichael** [comment]
We can add some better build tags around this.  I think we already have some things that check for cgo or !cgo

**Comment by mlaventure** [comment]
Just assume it's 64bit for now I guess?

Seeing a better way to handle this part @cyphar ? I'm open to all suggestions  :)

**Comment by cyphar** [comment]
`CHAR_BIT * sizeof(long int)` is okay but there's issues about running a 32-bit program on a 64-bit host. So we'll probably have to do something more clever than that.

**Comment by vishh** [comment]
@cyphar 
Assuming that kmem setting logic [here](https://github.com/opencontainers/runc/blob/88bb59e35f71bffa94b39722c0faf61e411c83bf/libcontainer/cgroups/fs/memory.go#L66) is the only user of `GetLongBit()`, can we have the kmem logic compare current memcg kmem limits with that of the root `/` container? Is runc expected to work within a container where not all cgroups are visible?

**Comment by dubstack** [comment]
I suggest that we modify the logic for setting the kmem.limit_on_bytes to be more deterministic and
remove the kmemInitialized check.

If kernel memory is enabled we will set the kmem.limit_in_bytes to -1 when the cgroup is created ie. in apply(). This will enable kernel memory accounting on the cgroup ( http://lxr.free-electrons.com/source/Documentation/cgroups/memory.txt?v=3.9#L284 ) Once accounting is enabled we should be able to set the kernel memory limit to the required value in Set().

The only scenario in which this would fail is if someone straight out calls Set() on a cgroup which was not created by libcontainer, in which case, if there are already process attached to the cgroup setting kernel memory limit would return an Ebusy signal and we can error out.

**Comment by vishh** [comment]
I don't think the APIs are expected to handle cgroups that are created out
of band. Setting `-1` will only enable accounting and will not affect
containers in any way. So +1 for setting `-1` as
`memory.kmem.limit_in_bytes` by default. if kmem is enabled.

**Comment by dubstack** [comment]
Re, above, In apply() just setting -1 won't limit the cgroup, so we will have to limit the kernel memory by writing some arbitrary limiting value(say 1) and then write -1 to the kmem.limit file. Both the writes would be done one after the other in Apply()

**Comment by hqhq** [comment]
The double write seems a doable solution, but that would loose the accurate error message on kernel memory setting failure.
@vishh Why didn't you compare current memcg kmem limits with that of the root `/` container as you suggested in https://github.com/opencontainers/runc/issues/841#issuecomment-230662085 ? I think that's a better idea.

**Comment by vishh** [comment]
@hqhq Comparing limits of parent requires access to parent limits and that is a lot more tedious. Essentially what we are trying to accomplish is that of prompting users whenever they fail to set kmem limits prior to moving processed into cgroups. 

I suggested the applying dummy limits approach to @dubstack to make the API more deterministic. Can you expand a bit more on what error message would be lost by following this method? What this approach does is that of enabling kmem accounting as part of cgroups creation. It is upto the user to apply kmem limits.

**Comment by dubstack** [comment]
From what I understand @hqhq 's concern is that we wont be able to prompt the user as to why exactly the kmem set() failed as he would just get a resource busy error. 

But as I and @vishh have already pointed out this would be a part of the cgroup creation and it would only error out if a user tries applying kmem limits to a cgroup which was not created by libcontainer.

**Comment by mlaventure** [comment]
@dubstack but that's a scenario that runc support, and it may make sense in some environment.

Can we just have an error message  on `EBUSY` akin to `can not override limit for cgroup %s, currently sets sets to %d: device or resource busy` on your PR?

That way the user would know they either need to not join that cgroup or remove the limit from their conf and accept the one currently enforced.

**Comment by dubstack** [comment]
@mlaventure I totally agree with you. I looked into how golang handles syscalls and I found out that error numbers are not floated back to the calls instead it returns the error string. So we can't find the EBUSY error from the returned error.

The less elegant solution would be to compare the "Resource Busy" error string but again thats difficult as the error string is not uniformly defined across different architectures.

**Comment by mlaventure** [comment]
@dubstack `WriteFile` returns an `os.PathError` in that case normal, which contains a `Err` field with the error number. If the type assertion on the returned error succeed on that type, that could do the trick.

WDYT?

**Comment by dubstack** [comment]
I am not completely sure if understood you correctly, but the problem is that the type assertion check would pass everytime a os.PathError is returned which can happen due to multiple reasons and not necessarily EBUSY. 

Please correct me if I am wrong but we would need some way of accessing the EBUSY error number which as far as I know is not possible this way. Or maybe I din't get the trick .

**Comment by mlaventure** [comment]
@dubstack  It's not very pretty, but something like:
```
if pe, ok := err.(*os.PathError); ok {
	if ern, ok := pe.Err.(syscall.Errno); ok && ern == syscall.EBUSY {
		fmt.Println("Special error message goes here")
	}
}
```

should work.

**Comment by dubstack** [comment]
@mlaventure Oh I should have been a bit more specific in the last comment, but the problem with what you are suggesting is that the error being returned by ioutil.WriteFile() is of type `error`. And when the resource busy error is encountered we return `&PathError()`, So you cannot access the field `Err` from the returned pointer PathError. Because the interface `error` type is being returned and we can't access the field of the struct PathError which is being returned.

**Comment by mlaventure** [comment]
@dubstack   This works for me though:

```go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"syscall"
)

func main() {
	f, err := os.OpenFile("toto", os.O_CREATE|os.O_TRUNC|os.O_RDONLY, 0444)
	if err != nil {
		fmt.Println(err)
		return
	}
	f.Close()

	e := ioutil.WriteFile("toto", []byte("ha!"), 0444)

	if val, ok := e.(*os.PathError); ok {
		if v, ok := val.Err.(syscall.Errno); ok {
			if v == syscall.EACCES {
				fmt.Println("SUCCESS")
			}
		}
		return
	}
	fmt.Printf("FAILURE")

}
```

`WriteFile` will just return the error from `File.Write` when it occurs, which implements the `error interface` in which case the assertion will succeed

**Comment by cyphar** [comment]
@dubstack You can use a type assertion with go to essentially "cast" the interface to the type `PathError`. See @mlaventure's examples.

**Comment by dubstack** [comment]
@mlaventure @cyphar Thanks for pointing that out. I wasn't aware that Go supports "type casting" in the following manner. I will update my patch with the suggested check.

Again, thanks a lot @mlaventure :)

**Comment by cyphar** [comment]
AFAICS this is fixed by #962 and #968. Closing.
